### ISSUE TITLE
Browser Extension - Enable web-clipping feature

### ISSUE THREAD

**Issue by RHuehne** [linkwarden/linkwarden#186]
Hi There,

I am just taking a first look at linkwarden. So far, it looks very nice.

However, just for my understanding: If I am using the browser extension, linkwarden server will take the clip of the target web page, not the extension locally from my local browser.

Therefore, if I want to store an article, that is behind paywall (paid), linkwarden will not be able to do it, as it cannot login and get behind paywall...? That would be a sad fact... 

Thanks

**Comment by crkos** [comment]
Yes that is correct, That is the way the extension is designed, we just send the server the request to save the links, so i would guess that paywalled content or any other content that is behind a login page pretty much will just receive some login page or any other page.

If i'm not mistaken linkwarden uses playwright to take screenshots(? i dont really know), so i think in the documentation there is a way to get authentication going (there is!) so maybe it can be solved in that way

**Comment by RHuehne** [comment]
sadly, in these days, this makes the whole story quite useless, as today, there is nearly no website that does not put any cookie-accept-popup on their pages. It even does not require a paywall. Not really helpful, when the content is greyed out before the cookie-crap-message is confirmed. 

The plugin should send local cookie and session infos to linkwarden, in order to have linkwarden accessing the page the same way as the user.... but I guess this ain't easy to achieve... 

I was hoping to find something similar to the web-clipper of evernote... but so far I did not see anyting like this on the open source market...

**Comment by shrippen** [comment]
the Wallabag browser extension has the ability to use the rendering of your browser to send it to wallabag. So that might be something to look into for you.

**Comment by RHuehne** [comment]
okay, thanks for input, will check that out. 
cheers

**Comment by mayzyo** [comment]
This was a huge surprise to me as well. It would make Linkwarden a killer app if it did that. Or even just an option to do the rendering from the extension and send the data to the backend as fallback. As @RHuehne said, most articles worth saving is behind a paywall.

**Comment by crkos** [comment]
Well basically this is possible, The api (browser) has the capability to capture the current page, The only thing is right now is that i dont know if i can send directly images/pdfs to linkwarden to save them in the links, I will talk @daniel31x13 about this

**Comment by daniel31x13** [comment]
@RHuehne the feature that you're mentioning can be _easily_ implemented, the main problem it occurs is we're basically enabling file uploads. And the problem with that would be things will be pivoting around "Files" instead of "Links" so we need to rename the project to something like "FileWarden"...

But anyway, I'll be changing the title of this issue to a feature request and see what you (and others) think as well.

**Comment by RHuehne** [comment]
> @RHuehne the feature that you're mentioning can be _easily_ implemented, the main problem it occurs is we're basically enabling file uploads. And the problem with that would be things will be pivoting around "Files" instead of "Links" so we need to rename the project to something like "FileWarden"...
> 
> But anyway, I'll be changing the title of this issue to a feature request and see what you (and others) think as well.
> 
> 👍 or 👎 this issue (top comment) to agree or disagree on adding this feature.

I see that point. Linkwarden covers different requirements. For myself I found wallabag (https://github.com/wallabag/wallabag), which covers my needs.

**Comment by fmbarina** [comment]
I don't understand the "FileWarden" metaphor.

From my understanding, Playwright (if that's what's being used) allows you to set cookies for a BrowserContext. A quick search reveals multiple extensions and methods available to extract cookies from a user's browser.

Would it not be possible to send the relevant page cookies to LinkWarden to be used in archival operations? Though not bulletproof, this may even alleviate issues such as #138.

Whether they are sent automatically during link bookmark creation or the user has to manually upload a `cookies.txt` file in the app interface seems to be a design and implementation decision, but even in the latter case it's rather extreme to claim this as pivoting to file uploads. I'm unfamiliar with Evernote's web-clipper, so apologies if I'm lacking some crucial context. This is how I understood RHuehne's request.

I myself have already stumbled into multiple websites where the archive functionality is severely impacted by banners, popups, paywalls, and even redirects.

I'm enjoying LW so far, but this issue sadly blocked a lot of what I intended to use it for.

**Comment by daniel31x13** [comment]
The web-clipping is already part of the roadmap, forgot to change the status of this issue.

Right now the plan is to first make linkwarden accept images/PDFs (which is already implemented inside the dev branch, just not yet released).

We’re planning to release all the new features together soon.

**Comment by crkos** [comment]
yup, i'm working on it, i'm just slow cuz i have heavy load at work right now.

**Comment by fmbarina** [comment]
So, if I'm reading this right, the web clipping functionality should work by uploading the saved clip/file from the user's browser where it was just generated/captured, through the use of the extension or something similar.

I'll understand if this isn't a use-case you've considered or ever intended to support, but how would that work for users of the extension on Android Firefox?

Archives being generated on the server means they're always the same format (e.g. screen shape). If the client starts generating and uploading content directly, wouldn't that mean that web clips will vary in format depending on where the user saved them? Assuming they even work on mobile, that is ;^)

**Comment by huyz** [comment]
@fmbarina Don't quote me, but I think the Evernote assumes a certain screen width when clipping a "full page" or "article". As for "Simplified Article" ("readable" version), then it doesn't matter.

Btw, I hope web-clipping functionality would support at least both these modes as one works better than the other depending on the web site. (Evernote also supports "screenshot", "bookmark", "multi-select", and an "article" mode that conveniently allows one to correct the selection of the proper `<div>`, but those could come later)

**Comment by DavidMoffatt-web** [comment]
Glad to see this is on the roadmap; this is the only feature I believe LinkWarden is missing to compete with some of the larger players in the space.

**Comment by Phlogi** [comment]
Please double check the implementation of the web clipping that readeck has done. It's well implemented. 
https://codeberg.org/readeck/browser-extension

**Comment by Phlogi** [comment]
Does the linkwarden API support sending actual HTML content to the backend already?

**Comment by daniel31x13** [comment]
You can now send the full page to Linkwarden using the browser extension.

**Comment by huyz** [comment]
@daniel31x13 Hmmm, I have the extension v1.2.8 and I only see the option to upload an image, not the full page.

**Comment by daniel31x13** [comment]
@huyz By "full page" I meant full page screenshot.

**Comment by huyz** [comment]
Ok so that's the equivalent of Evernote's screenshot. But I'm guessing most people would want a web clipping of HTML, e.g. so that we can select and copy text from the clip in the future.
I guess we'd need to open a new issue for that
